1-Parallelize-your-python-code_Futures_API.ipynb

```diff
@@ -0,0 +1,893 @@
+    "    \n",
+    "        best_answer_author_obj = answ.find(\"span\", itemprop=\"name\")\n",
+    "        \n",
+    "        if best_answer_author_obj:\n",
+    "            best_answer_author = best_answer_author_obj.text\n",
+    "        else:\n",
+    "            best_answer_author = \"comunity_post\"\n",
+    "\n",
+    "        post_info[\"best_answer_usrname\"] = best_answer_author\n",
+    "    else:\n",
+    "        post_info[\"best_answer_votes\"] = 0\n",
+    "        post_info[\"best_answer_usrname\"] = \"no-answer\"\n",
+    "\n",
+    "    \n",
+    "    return post_info"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7ea45c41-734c-4dc0-bc64-cd3b401eb9a6",
+   "metadata": {},
+   "source": [
+    "## Serial"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2402f42a-e528-45dd-bce8-0b3bc64022f7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "df_list =[]\n",
+    "for page_num in range(1, 3): #more than 2 pages and get trhottling issues\n",
+    "    page_html = get_page_html_links(page_num)\n",
+    "    posts_links = get_post_links_per_page(page_html)\n",
+    "    list_post_data = []\n",
+    "    \n",
+    "    for link in posts_links:\n",
+    "        p_data = get_data(link)\n",
+    "        list_post_data.append(p_data)\n",
+    "\n",
+    "    df = pd.DataFrame(list_post_data)\n",
+    "    df_list.append(df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8a8a0cf1-4b0c-48d9-a037-d43306ff510c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_list[0].head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "19cc84dc-bd05-475a-be57-398e0fb391af",
+   "metadata": {},
+   "source": [
+    "## Parallel"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9f380cdb-2c52-48f5-b1b7-32549dcc704f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from dask.distributed import wait, as_completed"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7870844e-5711-45fc-8b8c-81a397be48e5",
+   "metadata": {},
+   "source": [
+    "### Get pages and links of posts"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "41423031-c1ac-4070-b34d-c069a8cd1b49",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "pages_futures = client.map(get_page_html_links, range(1,3))\n",
+    "wait(pages_futures)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e7a37554-b0a5-4df0-bd73-2f2f5c9f5c30",
+   "metadata": {},
+   "source": [
+    "**`wait()`**\n",
+    "\n",
+    "Notice that here we used `wait()`, you can wait on a future or collection of futures using the `wait` function, which blocks until all futures are finished or have erred. This is useful when you need the all the futures to be completed to proceed with your computations. \n",
+    "\n",
```

Comment:
If the results are small you might also use `results = [future.result() for future in futures]`

This is computationally worse, but it does mean that you can introduce them to fewer pieces of API.  (less for them to get confused by)